README.md
# CustomProgressButton 

`CustomProgressButton` is a customizable button widget for Flutter that allows displaying different states based on the `ButtonState` enum. It supports a loading state with an optional progress indicator. The appearance and behavior of the button can be customized using various properties.

## Installation

Add the following dependency to your `pubspec.yaml` file:

```yaml
dependencies:
  custom_progress_button : ^2.0.1
```

Then, run `flutter pub get` to install the package.

## Usage

Import the package in your Dart file:

```dart
import 'package:progress_button/progress_button.dart';
```

### Basic Usage

To use the `ProgressButton`, provide a map of widgets for different button states and a map of colors for different states. Then, set the `state` property to control the current state of the button.

```dart
CustomProgressButton(
  stateWidgets: {
    ButtonState.idle: Text('Submit'), // Idle state widget
    ButtonState.loading: CircularProgressIndicator(), // Loading state widget
    ButtonState.success: Icon(Icons.check), // Success state widget
    ButtonState.fail: Icon(Icons.close), // Fail state widget
  },
  stateColors: {
    ButtonState.idle: Colors.blue, // Idle state color
    ButtonState.loading: Colors.grey, // Loading state color
    ButtonState.success: Colors.green, // Success state color
    ButtonState.fail: Colors.red, // Fail state color
  },
  state: ButtonState.idle, // Initial state
  onPressed: () {
    // Handle button press
  },
),
```

### Advanced Usage

#### Customizing Button Properties

You can customize various properties of the `ProgressButton`:

```dart
CustomProgressButton(
  // ... other properties
  minWidth: 200.0, // Minimum width of the button
  maxWidth: 400.0, // Maximum width of the button
  radius: 20.0, // Corner radius of the button
  height: 48.0, // Height of the button
  progressIndicatorSize: 20.0, // Size of the progress indicator
  progressAlignment: MainAxisAlignment.center, // Alignment of the progress indicator
  padding: EdgeInsets.symmetric(horizontal: 16.0), // Padding around the button's child
  minWidthStates: [ButtonState.loading], // List of states that should use the minimum width
  buttonShapeEnum: ButtonShapeEnum.elevated, // Shape of the button (elevated, outline, or flat)
  elevation: 4.0, // Elevation of the button (used only for elevated and outlined shapes)
  inLineBackgroundColor: Colors.white, // Background color for loading and success states
  enable: true, // Enable or disable the button
),
```

#### Using Icon Buttons

You can use the `ProgressButton.icon` constructor to create buttons with icon buttons for different states:

```dart
CustomProgressButton.icon(
  iconButtons: {
    ButtonState.idle: CustomIconButton(
      text: 'Submit',
      icon: Icon(Icons.send),
      color: Colors.blue,
    ),
    ButtonState.loading: CustomIconButton(
      icon: CircularProgressIndicator(),
      color: Colors.grey,
    ),
    ButtonState.success: CustomIconButton(
      icon: Icon(Icons.check),
      color: Colors.green,
    ),
    ButtonState.fail: CustomIconButton(
      icon: Icon(Icons.close),
      color: Colors.red,
    ),
  },
  onPressed: () {
    // Handle button press
  },
),
```

## CustomIconButton Class

The `CustomIconButton` class represents an icon button with optional text and color. You can use this class to customize the appearance of the `ProgressButton` icon buttons.

```dart
class CustomIconButton {
  final String? text;
  final Widget? icon;
  final Color color;

  const CustomIconButton({
    this.text,
    this.icon,
    required this.color,
  });
}
```

## Utility Functions

The `CustomProgressButton` package also provides utility functions to help build widgets containing icons, text, and gaps with specified styles:

- `buildChildWithIcon`: A function to build a widget containing an icon and optional text with a specified gap.
- `buildChildWithIC`: A function to build a widget containing text and an icon with a specified gap.
- `buildText`: A function to build a text widget with the provided text and text style.

## Feedback and Contributions

We welcome feedback and contributions! If you encounter any issues or have suggestions for improvements, please [file an issue](https://github.com/georgesamirmansour/custom_progress_button/issues) on GitHub.

If you want to contribute to the project, feel free to open a pull request. We appreciate your help in making `ProgressButton` even better!

## License

This project is licensed under the [MIT License](https://github.com/georgesamirmansour/custom_progress_button/blob/master/LICENSE).

CHANGELOG.md
## Changelog

### Version 2.0.1
- update README

### Version 2.0.0
- **Breaking Change**: Renamed the package from `progress_button` to `custom_progress_button` to avoid naming conflicts.
- Added support for null-safety.
- Updated the `custom_icon_button` dependency to version 2.0.0, which also supports null-safety.
- Updated Flutter SDK version to support null-safety.
- Updated documentation and examples for clarity and accuracy.
- Improved overall performance and stability.

### Version 1.2.0
- Added support for customizing the button's corner radius.
- Added new properties to customize the button's appearance:
    - `elevation`: The elevation of the button for elevated and outlined shapes.
    - `inLineBackgroundColor`: The background color of the button when it's in the loading or success state.
    - `enable`: If false, the button will be disabled and not respond to touch events.
    - `progressAlignment`: The alignment of the progress indicator when `progressWidget` is not provided.
    - `minWidthStates`: The list of states that should use the minimum width of the button.
    - `buttonShapeEnum`: The shape of the button (elevated, outline, or flat).
- Fixed a bug where the button's appearance would not update correctly when the state changes.

### Version 1.1.0
- Added support for creating buttons with icon buttons for different states using the `ProgressButton.icon` constructor.
- Introduced the `CustomIconButton` class to represent an icon button with optional text and color.
- Added utility functions to help build widgets containing icons, text, and gaps with specified styles.
- Improved the README with comprehensive usage instructions and examples.

### Version 1.0.0
- Initial release of the `progress_button` package.
- Provides a customizable button widget that shows different states based on the `ButtonState` enum.
- Supports a loading state with an optional progress indicator.
- Allows customization of appearance and behavior using various properties.

lib/custom_progress.dart
import 'package:flutter/material.dart';
import 'custom_icon_button.dart';

enum ButtonState { idle, loading, success, fail }
enum ButtonShapeEnum { elevated, outline, flat }

class ProgressButton extends StatefulWidget {
  final Map<ButtonState, Widget> stateWidgets;
  final Map<ButtonState, Color> stateColors;
  final VoidCallback onPressed;
  final Function? onAnimationEnd;
  late ButtonState state;
  late final double minWidth;
  late final double maxWidth;
  final double? radius;
  late final double height;
  final Widget? progressWidget;
  late final double progressIndicatorSize;
  late final MainAxisAlignment progressAlignment;
  late final EdgeInsets padding;
  late final List<ButtonState> minWidthStates;
  late final ButtonShapeEnum buttonShapeEnum;
  late final double elevation;
  late final Color inLineBackgroundColor;
  late final bool enable;

  ProgressButton(
      {Key? key,
      required this.stateWidgets,
      required this.stateColors,
      this.state = ButtonState.idle,
      required this.onPressed,
      this.onAnimationEnd,
      this.minWidth = 200.0,
      this.maxWidth = 400.0,
      this.radius,
      this.height = 36.0,
      this.progressIndicatorSize = 10.0,
      this.progressWidget,
      this.progressAlignment = MainAxisAlignment.spaceEvenly,
      this.padding = EdgeInsets.zero,
      this.minWidthStates = const <ButtonState>[ButtonState.loading],
      this.buttonShapeEnum = ButtonShapeEnum.elevated,
      this.elevation = 8.0,
      this.enable = true,
      this.inLineBackgroundColor = Colors.white})
      : assert(
          stateWidgets.keys.toSet().containsAll(ButtonState.values.toSet()),
          'Must be non-null widgets provided in map of stateWidgets. Missing keys => ${ButtonState.values.toSet().difference(stateWidgets.keys.toSet())}',
        ),
        assert(
          stateColors.keys.toSet().containsAll(ButtonState.values.toSet()),
          'Must be non-null widgetds provided in map of stateWidgets. Missing keys => ${ButtonState.values.toSet().difference(stateColors.keys.toSet())}',
        ),
        super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _ProgressButtonState();
  }

  factory ProgressButton.icon({
    required Map<ButtonState, CustomIconButton> iconButtons,
    required VoidCallback onPressed,
    ButtonState state = ButtonState.idle,
    Function? onAnimationEnd,
    maxWidth: 170.0,
    minWidth: 58.0,
    height: 53.0,
    radius: 100.0,
    progressIndicatorSize: 35.0,
    double iconPadding: 4.0,
    TextStyle? textStyle,
    ButtonShapeEnum buttonShapeEnum = ButtonShapeEnum.elevated,
    double elevation = 8.0,
    Color inLineBackgroundColor = Colors.white,
    bool enable = true,
    Widget? progressWidget,
    MainAxisAlignment? progressIndicatorAlignment,
    EdgeInsets padding = EdgeInsets.zero,
    List<ButtonState> minWidthStates = const <ButtonState>[ButtonState.loading],
  }) {
    assert(
      iconButtons != null &&
          iconButtons.keys.toSet().containsAll(ButtonState.values.toSet()),
      'Must be non-null widgets provided in map of stateWidgets. Missing keys => ${ButtonState.values.toSet().difference(iconButtons.keys.toSet())}',
    );

    if (textStyle == null) {
      textStyle = TextStyle(color: Colors.white, fontWeight: FontWeight.w500);
    }

    Map<ButtonState, Widget> stateWidgets = {
      ButtonState.idle: buildChildWithIcon(
          iconButtons[ButtonState.idle]!, iconPadding, textStyle),
      ButtonState.loading: Column(),
      ButtonState.fail: buildChildWithIcon(
          iconButtons[ButtonState.fail]!, iconPadding, textStyle),
      ButtonState.success: buildChildWithIcon(
          iconButtons[ButtonState.success]!, iconPadding, textStyle)
    };

    Map<ButtonState, Color> stateColors = {
      ButtonState.idle: iconButtons[ButtonState.idle]!.color,
      ButtonState.loading: iconButtons[ButtonState.loading]!.color,
      ButtonState.fail: iconButtons[ButtonState.fail]!.color,
      ButtonState.success: iconButtons[ButtonState.success]!.color,
    };

    return ProgressButton(
      stateWidgets: stateWidgets,
      stateColors: stateColors,
      state: state,
      onPressed: onPressed,
      onAnimationEnd: onAnimationEnd,
      maxWidth: maxWidth,
      minWidth: minWidth,
      radius: radius,
      height: height,
      progressIndicatorSize: progressIndicatorSize,
      progressAlignment: MainAxisAlignment.center,
      progressWidget: progressWidget,
      minWidthStates: minWidthStates,
      elevation: elevation,
      padding: padding,
      buttonShapeEnum: buttonShapeEnum,
      inLineBackgroundColor: inLineBackgroundColor,
      enable: enable,
    );
  }
}

class _ProgressButtonState extends State<ProgressButton>
    with TickerProviderStateMixin {
  AnimationController? colorAnimationController;
  Animation<Color?>? colorAnimation;
  double? width;
  Duration animationDuration = Duration(milliseconds: 500);
  Widget? progressWidget;

  void startAnimations(ButtonState? oldState, ButtonState? newState) {
    Color? begin = widget.stateColors[oldState!];
    Color? end = widget.stateColors[newState!];
    if (widget.minWidthStates.contains(newState)) {
      width = widget.minWidth;
    } else {
      width = widget.maxWidth;
    }
    colorAnimation = _colorAnimation(begin, end);
    colorAnimationController!.forward();
  }

  Animation<Color?> _colorAnimation(Color? begin, Color? end) =>
      ColorTween(begin: begin, end: end).animate(CurvedAnimation(
        parent: colorAnimationController!,
        curve: Interval(
          0,
          1,
          curve: Curves.easeIn,
        ),
      ));

  Color? get backgroundColor => colorAnimation == null
      ? widget.stateColors[widget.state]
      : colorAnimation!.value ?? widget.stateColors[widget.state];

  @override
  void initState() {
    super.initState();
    width = widget.maxWidth;
    colorAnimationController = _animationController;
    colorAnimationController!.addStatusListener((status) {
      if (widget.onAnimationEnd != null) {
        widget.onAnimationEnd!(status, widget.state);
      }
    });

    progressWidget = _progress;
  }

  AnimationController get _animationController =>
      AnimationController(duration: animationDuration, vsync: this);

  Widget get _progress =>
      widget.progressWidget ??
      CircularProgressIndicator(
          backgroundColor: widget.stateColors[widget.state],
          valueColor: AlwaysStoppedAnimation<Color>(Colors.white));

  @override
  void dispose() {
    colorAnimationController!.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(ProgressButton oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.state != widget.state) {
      colorAnimationController?.reset();
      startAnimations(oldWidget.state, widget.state);
    }
  }

  Widget getButtonChild(bool visibility) {
    Widget? buttonChild = widget.stateWidgets[widget.state];
    if (widget.state == ButtonState.loading) {
      return Row(
        mainAxisAlignment: widget.progressAlignment,
        children: <Widget>[
          SizedBox(
            child: progressWidget,
            width: widget.minWidth / 3,
            height: widget.height > 55
                ? widget.height / 2
                : widget.progressIndicatorSize,
          ),
          buttonChild ?? Container(),
        ],
      );
    }
    return _animatedOpacity(visibility, buttonChild);
  }

  AnimatedOpacity _animatedOpacity(bool visibility, Widget? child) =>
      AnimatedOpacity(
          opacity: visibility ? 1.0 : 0.0,
          duration: Duration(milliseconds: 250),
          child: child);

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: colorAnimationController!,
      builder: (context, child) {
        return AnimatedContainer(
            width: width,
            height: widget.height,
            duration: animationDuration,
            child: MaterialButton(
              padding: widget.padding,
              shape: buttonShape,
              color: _backgroundColor,
              elevation: _buttonElevation,
              onPressed: widget.enable ? widget.onPressed : null,
              child: _buttonChild,
            ));
      },
    );
  }

  Widget get _buttonChild => getButtonChild(
      colorAnimation == null ? true : colorAnimation!.isCompleted);

  double get _buttonElevation =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.outline
          ? widget.elevation
          : 0.0;

  RoundedRectangleBorder get buttonShape =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.flat
          ? _roundedRectangleBorderElevated
          : _roundedRectangleBorderOutlined;

  Color get _backgroundColor =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.flat
          ? backgroundColor!
          : widget.inLineBackgroundColor;

  RoundedRectangleBorder get _roundedRectangleBorderElevated =>
      RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(
              widget.radius == null ? widget.height / 2 : widget.radius!),
          side: BorderSide(color: Colors.transparent, width: 0));

  RoundedRectangleBorder get _roundedRectangleBorderOutlined =>
      RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(
              widget.radius == null ? widget.height / 2 : widget.radius!),
          side: BorderSide(color: backgroundColor!, width: 1));
}


lib/custom_icon_button.dart
import 'package:flutter/material.dart';

// A custom class representing an icon button with optional text and color.
class CustomIconButton {
  // The optional text to display alongside the icon.
  final String? text;

  // The optional icon widget to display.
  final Widget? icon;

  // The color of the icon button.
  final Color color;

  const CustomIconButton({
    this.text,
    this.icon,
    required this.color,
  });
}

/// A utility function to build a widget containing an icon and optional text with a specified gap.
///
/// Parameters:
/// - [iconButton]: The custom icon button that contains the icon, optional text, and color.
/// - [iconPadding]: The padding between the icon and the text, if text is provided.
/// - [textStyle]: The style for the text, if text is provided.
///
/// Returns:
/// A widget containing the icon and optional text wrapped with appropriate padding.
Widget buildChildWithIcon(
    CustomIconButton iconButton, double iconPadding, TextStyle textStyle) {
  return buildChildWithIC(
      iconButton.text, iconButton.icon, iconPadding, textStyle);
}

/// A utility function to build a widget containing text and an icon with a specified gap.
///
/// Parameters:
/// - [text]: The text to display.
/// - [icon]: The icon widget to display.
/// - [gap]: The padding between the text and the icon.
/// - [textStyle]: The style for the text.
///
/// Returns:
/// A widget containing the provided text and icon wrapped with appropriate padding.
Widget buildChildWithIC(
    String? text, Widget? icon, double gap, TextStyle textStyle) {
  var children = <Widget>[];
  children.add(icon ?? Container());
  if (text != null) {
    children.add(Padding(padding: EdgeInsets.all(gap)));
    children.add(buildText(text, textStyle));
  }

  return Wrap(
    direction: Axis.horizontal,
    crossAxisAlignment: WrapCrossAlignment.center,
    children: children,
  );
}

/// A utility function to build a text widget with the provided text and text style.
///
/// Parameters:
/// - [text]: The text to display.
/// - [style]: The style for the text.
///
/// Returns:
/// A text widget displaying the provided text with the specified style.
Widget buildText(String text, TextStyle style) {
  return Text(text, style: style);
}



lib/custom_progress_button.dart
import 'package:flutter/material.dart';
import 'custom_icon_button.dart';

// Enum to represent the different states of the button
enum ButtonState { idle, loading, success, fail }

// Enum to represent different button shapes
enum ButtonShapeEnum { elevated, outline, flat }

/// A customizable button that shows different states based on the [ButtonState].
/// It supports a loading state with an optional progress indicator.
/// The appearance and behavior of the button can be customized using various properties.
class CustomProgressButton extends StatefulWidget {
  // A map containing widgets to display for different button states.
  final Map<ButtonState, Widget> stateWidgets;

  // A map containing colors for different button states.
  final Map<ButtonState, Color> stateColors;

  // The function to be called when the button is pressed.
  final VoidCallback onPressed;

  // A function that is called when the animation ends.
  final Function? onAnimationEnd;

  // The initial state of the button. Defaults to [ButtonState.idle].
  late final ButtonState state;

  // The minimum width of the button. Defaults to 200.0.
  late final double minWidth;

  // The maximum width of the button. Defaults to 400.0.
  late final double maxWidth;

  // The corner radius of the button. If null, it will be set to half of the button's height.
  final double? radius;

  // The height of the button. Defaults to 36.0.
  late final double height;

  // The optional widget to display while the button is in the loading state.
  final Widget? progressWidget;

  // The size of the progress indicator. Defaults to 10.0.
  late final double progressIndicatorSize;

  // The alignment of the progress indicator when [progressWidget] is not provided.
  // Defaults to [MainAxisAlignment.spaceEvenly].
  late final MainAxisAlignment progressAlignment;

  // Padding around the button's child (content).
  late final EdgeInsets padding;

  // The list of states that should use the minimum width of the button.
  late final List<ButtonState> minWidthStates;

  // The shape of the button. Defaults to [ButtonShapeEnum.elevated].
  late final ButtonShapeEnum buttonShapeEnum;

  // The elevation of the button. Used only for [ButtonShapeEnum.elevated] and [ButtonShapeEnum.outline].
  late final double elevation;

  // The background color of the button when it's in the loading or success state.
  late final Color inLineBackgroundColor;

  // If false, the button will be disabled and not respond to touch events. Defaults to true.
  late final bool enable;

  CustomProgressButton({
    Key? key,
    required this.stateWidgets,
    required this.stateColors,
    this.state = ButtonState.idle,
    required this.onPressed,
    this.onAnimationEnd,
    this.minWidth = 200.0,
    this.maxWidth = 400.0,
    this.radius,
    this.height = 40.0,
    this.progressIndicatorSize = 10.0,
    this.progressWidget,
    this.progressAlignment = MainAxisAlignment.spaceEvenly,
    this.padding = EdgeInsets.zero,
    this.minWidthStates = const <ButtonState>[ButtonState.loading],
    this.buttonShapeEnum = ButtonShapeEnum.elevated,
    this.elevation = 8.0,
    this.enable = true,
    this.inLineBackgroundColor = Colors.white,
  })  : assert(
          stateWidgets.keys.toSet().containsAll(ButtonState.values.toSet()),
          'Must provide non-null widgets for all ButtonState values. Missing keys: ${ButtonState.values.toSet().difference(stateWidgets.keys.toSet())}',
        ),
        assert(
          stateColors.keys.toSet().containsAll(ButtonState.values.toSet()),
          'Must provide non-null colors for all ButtonState values. Missing keys: ${ButtonState.values.toSet().difference(stateColors.keys.toSet())}',
        ),
        super(key: key);

  // A factory constructor to create a [ProgressButton] with icon buttons for different states.
  factory CustomProgressButton.icon({
    required Map<ButtonState, CustomIconButton> iconButtons,
    required VoidCallback onPressed,
    ButtonState state = ButtonState.idle,
    Function? onAnimationEnd,
    maxWidth = 170.0,
    minWidth = 58.0,
    height = 53.0,
    radius = 100.0,
    progressIndicatorSize = 35.0,
    double iconPadding = 4.0,
    TextStyle? textStyle,
    ButtonShapeEnum buttonShapeEnum = ButtonShapeEnum.elevated,
    double elevation = 8.0,
    Color inLineBackgroundColor = Colors.white,
    bool enable = true,
    Widget? progressWidget,
    MainAxisAlignment? progressIndicatorAlignment,
    EdgeInsets padding = EdgeInsets.zero,
    List<ButtonState> minWidthStates = const <ButtonState>[ButtonState.loading],
  }) {
    assert(
      iconButtons.isNotEmpty &&
          iconButtons.keys.toSet().containsAll(ButtonState.values.toSet()),
      'Must provide non-null widgets for all ButtonState values. Missing keys: ${ButtonState.values.toSet().difference(iconButtons.keys.toSet())}',
    );

    if (textStyle == null) {
      textStyle = TextStyle(color: Colors.white, fontWeight: FontWeight.w500);
    }

    // Create a map of widgets for different states.
    Map<ButtonState, Widget> stateWidgets = {
      ButtonState.idle: buildChildWithIcon(
          iconButtons[ButtonState.idle]!, iconPadding, textStyle),
      ButtonState.loading: Column(),
      ButtonState.fail: buildChildWithIcon(
          iconButtons[ButtonState.fail]!, iconPadding, textStyle),
      ButtonState.success: buildChildWithIcon(
          iconButtons[ButtonState.success]!, iconPadding, textStyle),
    };

    // Create a map of colors for different states.
    Map<ButtonState, Color> stateColors = {
      ButtonState.idle: iconButtons[ButtonState.idle]!.color,
      ButtonState.loading: iconButtons[ButtonState.loading]!.color,
      ButtonState.fail: iconButtons[ButtonState.fail]!.color,
      ButtonState.success: iconButtons[ButtonState.success]!.color,
    };

    // Return a new [ProgressButton] instance with the configured properties.
    return CustomProgressButton(
      stateWidgets: stateWidgets,
      stateColors: stateColors,
      state: state,
      onPressed: onPressed,
      onAnimationEnd: onAnimationEnd,
      maxWidth: maxWidth,
      minWidth: minWidth,
      radius: radius,
      height: height,
      progressIndicatorSize: progressIndicatorSize,
      progressAlignment: MainAxisAlignment.center,
      progressWidget: progressWidget,
      minWidthStates: minWidthStates,
      elevation: elevation,
      padding: padding,
      buttonShapeEnum: buttonShapeEnum,
      inLineBackgroundColor: inLineBackgroundColor,
      enable: enable,
    );
  }

  @override
  State<StatefulWidget> createState() {
    return _CustomProgressButtonState();
  }
}

// The private state class for [ProgressButton]
class _CustomProgressButtonState extends State<CustomProgressButton>
    with TickerProviderStateMixin {
  AnimationController? colorAnimationController;
  Animation<Color?>? colorAnimation;
  double? width;
  Duration animationDuration = Duration(milliseconds: 500);
  Widget? progressWidget;

  // Start color animations when the state changes
  void startAnimations(ButtonState? oldState, ButtonState? newState) {
    Color? begin = widget.stateColors[oldState!];
    Color? end = widget.stateColors[newState!];
    if (widget.minWidthStates.contains(newState)) {
      width = widget.minWidth;
    } else {
      width = widget.maxWidth;
    }
    colorAnimation = _colorAnimation(begin, end);
    colorAnimationController!.forward();
  }

  // Create a color animation between the given [begin] and [end] colors
  Animation<Color?> _colorAnimation(Color? begin, Color? end) =>
      ColorTween(begin: begin, end: end).animate(CurvedAnimation(
        parent: colorAnimationController!,
        curve: Interval(
          0,
          1,
          curve: Curves.easeIn,
        ),
      ));

  // Get the background color based on the current animation value or the current state
  Color? get backgroundColor => colorAnimation == null
      ? widget.stateColors[widget.state]
      : colorAnimation!.value ?? widget.stateColors[widget.state];

  @override
  void initState() {
    super.initState();
    width = widget.maxWidth;
    colorAnimationController = _animationController;
    colorAnimationController!.addStatusListener((status) {
      if (widget.onAnimationEnd != null) {
        widget.onAnimationEnd!(status, widget.state);
      }
    });

    progressWidget = _progress;
  }

  // Get a new animation controller
  AnimationController get _animationController =>
      AnimationController(duration: animationDuration, vsync: this);

  // Get the progress widget or a default CircularProgressIndicator if not provided
  Widget get _progress =>
      widget.progressWidget ??
      CircularProgressIndicator(
        backgroundColor: widget.stateColors[widget.state],
        valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
      );

  @override
  void dispose() {
    colorAnimationController!.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(CustomProgressButton oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.state != widget.state) {
      colorAnimationController?.reset();
      startAnimations(oldWidget.state, widget.state);
    }
  }

  // Get the button child based on whether it's in the loading state or not
  Widget getButtonChild(bool visibility) {
    Widget? buttonChild = widget.stateWidgets[widget.state];
    if (widget.state == ButtonState.loading) {
      return Row(
        mainAxisAlignment: widget.progressAlignment,
        children: <Widget>[
          SizedBox(
            child: progressWidget,
            width: widget.minWidth / 3,
            height: widget.height > 55
                ? widget.height / 2
                : widget.progressIndicatorSize,
          ),
          buttonChild ?? Container(),
        ],
      );
    }
    return _animatedOpacity(visibility, buttonChild);
  }

  // Get an animated opacity widget based on visibility and the child widget
  AnimatedOpacity _animatedOpacity(bool visibility, Widget? child) =>
      AnimatedOpacity(
        opacity: visibility ? 1.0 : 0.0,
        duration: Duration(milliseconds: 250),
        child: child,
      );

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: colorAnimationController!,
      builder: (context, child) {
        return AnimatedContainer(
          width: width,
          height: widget.height,
          duration: animationDuration,
          child: MaterialButton(
            padding: widget.padding,
            shape: buttonShape,
            color: _backgroundColor,
            elevation: _buttonElevation,
            onPressed: widget.enable ? widget.onPressed : null,
            child: _buttonChild,
          ),
        );
      },
    );
  }

  // Get the button child based on the current animation value
  Widget get _buttonChild => getButtonChild(
      colorAnimation == null ? true : colorAnimation!.isCompleted);

  // Get the elevation for elevated and outlined button shapes
  double get _buttonElevation =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.outline
          ? widget.elevation
          : 0.0;

  // Get the button shape based on the current button shape enum
  RoundedRectangleBorder get buttonShape =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.flat
          ? _roundedRectangleBorderElevated
          : _roundedRectangleBorderOutlined;

  // Get the background color for the outlined button shape
  Color get _backgroundColor =>
      widget.buttonShapeEnum == ButtonShapeEnum.elevated ||
              widget.buttonShapeEnum == ButtonShapeEnum.flat
          ? backgroundColor!
          : widget.inLineBackgroundColor;

  // Get the rounded rectangle border for the elevated button shape
  RoundedRectangleBorder get _roundedRectangleBorderElevated =>
      RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(
          widget.radius == null ? widget.height / 2 : widget.radius!,
        ),
        side: BorderSide(color: Colors.transparent, width: 0),
      );

  // Get the rounded rectangle border for the outlined button shape
  RoundedRectangleBorder get _roundedRectangleBorderOutlined =>
      RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(
          widget.radius == null ? widget.height / 2 : widget.radius!,
        ),
        side: BorderSide(color: backgroundColor!, width: 1),
      );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:custom_progress_button/custom_progress_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'progress button',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'progress button demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  late ButtonState buttonState;

  @override
  void initState() {
    buttonState = ButtonState.idle;
    super.initState();
  }


  void _incrementCounter() {
    Future.delayed(Duration(seconds: 2)).then((value) {
      setState(() {
        _counter++;
        if (buttonState == ButtonState.idle)
          buttonState = ButtonState.loading;
        else if (buttonState == ButtonState.loading)
          buttonState = ButtonState.fail;
        else if (buttonState == ButtonState.fail)
          buttonState = ButtonState.success;
        else if (buttonState == ButtonState.success)
          buttonState = ButtonState.idle;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
            SizedBox(
              height: 20,
            ),
            CustomProgressButton(
              // change button text depend on current state
              stateWidgets: {
                ButtonState.success: Text(
                  'Success',
                  style: TextStyle(fontSize: 20),
                ),
                ButtonState.fail: Text(
                  'Fail',
                  style: TextStyle(fontSize: 20),
                ),
                ButtonState.loading: Text(
                  'Loading',
                  style: TextStyle(fontSize: 20),
                ),
                ButtonState.idle: Text(
                  'Idle',
                  style: TextStyle(fontSize: 20),
                ),
              },
              // used in case of ButtonShapeEnum = ButtonShapeEnum.flat
              // this is line rounded to button
              inLineBackgroundColor: Colors.red,
              // change button color depending on current button state
              stateColors: {
                ButtonState.success: Colors.green,
                ButtonState.fail: Colors.redAccent,
                ButtonState.loading: Colors.red,
                ButtonState.idle: Colors.blue,
              },
              // called when animation of button ended
              onAnimationEnd: () {},
              // button radius
              radius: 100.0,
              // button height
              height: 70,
              // button shape enum as elevated, outline, flat
              buttonShapeEnum: ButtonShapeEnum.flat,
              // change default elevation of button,
              elevation: 10.0,
              // change alignment between progress and text of button in case of loading state
              progressAlignment: MainAxisAlignment.center,
              // disable/enable clicking on button
              enable: true,
              // button states are idle, loading, success, fail
              state: ButtonState.idle,
              // add padding
              padding: EdgeInsets.zero,
              // add max width to button
              maxWidth: 200.0,
              // default min width of button
              minWidth: 100.0,
              // progress size
              progressIndicatorSize: 10.0,
              // change default progress widget of button, default is
              progressWidget: CircularProgressIndicator(),
              // on pressed action
              onPressed: () {
                _incrementCounter();
              },
            )
          ],
        ),
      ),
    );
  }
}
